mpt_trie/src/trie_subsets.rs

```diff
@@ -307,13 +307,22 @@ fn mark_nodes_that_are_needed<N: PartialTrie>(
             return mark_nodes_that_are_needed(&mut children[nib as usize], curr_nibbles);
         }
         TrackedNodeIntern::Extension(child) => {
-            let nibbles = trie.info.get_nibbles_expected();
-            let r = curr_nibbles.pop_nibbles_front(nibbles.count);
+            // If we hit an extension node, we always must include it unless we exhausted
+            // our key.
+            if curr_nibbles.is_empty() {
+                return Ok(());
+            }
```

Comment:
I *think* that this actually should probably apply to leaf nodes as well. If we for example traverse down an extension or branch node and have no nibbles left in our key, I think this should be safe to say that we don't need the leaf?